### ISSUE TITLE
[Feature]: Integrate Triton MoE Kernel

### ISSUE THREAD

**Issue by zhuohan123** [vllm-project/vllm#16294]
### 🚀 The feature, motivation and pitch

Triton released a new MoE kernel: https://github.com/triton-lang/triton/pull/6429

Worth integrating into vLLM.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by aaron-ang** [comment]
hi, I'm interested in contributing but am new to the project. Which files or directories would you recommend me go over for reference?

**Comment by pylNeuralNet** [comment]
Hi, can the kernel be benchmarked on T4 gpu (supports FP16)? If nothing against, I would also like to help. Very interesting.

**Comment by sniper35** [comment]
@zhuohan123  I am looking it, could you assign it to me?

**Comment by pylNeuralNet** [comment]
Ok there are several kernels (matmul_ogs, SwiGLU, ...), and they need to be added with support for quantization and layer fusion. Of course, SwiGLU depends on the output gating scaling operation. 
How about we split the work and for example I work on SwiGLU and @sniper35 takes some next kernel? What do you guys think about it? Anyways I investigated it too, to learn about vLLM. Regardless, please add me to reviewers.

One thing that may be problematic is adapting fused_moe - I'm not sure how to handle this abstraction at this moment.
For the most part it uses cuda, like: `torch.ops._C.silu_and_mul(intermediate_cache2, intermediate_cache1.view(-1, N))`
I think that maybe something like:
`torch.ops.vllm.swiglu(intermediate_cache1.view(-1, N), out=intermediate_cache2)` and just copy the class there from triton? I don't think importing from Triton will be enough, but maybe I'm wrong and it actually can be as simple as import?

**Comment by zyongye** [comment]
@pylNeuralNet Hi, thanks for the message. I can join you all as well? From the triton PR, I only see matmul and SwiGLU, so I think that's the two big kernels we need (plus quantization support)? How about I take on matmul and @sniper35 can maybe take a look at quantization support? 

Again, feedback from the maintainer is always needed.

**Comment by simon-mo** [comment]
Hi everyone, thank you for your interested and contributions so far! To clarify, the task at hand is to benchmark the kernel by itself against the existing fused moe code path in vLLM, on a variety of models and devices. We expect something like the following to start, as example: "for Mixtral 8x7B on 8xH20, we see 12% throughput enhancement compare Triton with FusedMoe kernel for seqlen of 1024". 

Please feel free to work together.

**Comment by pylNeuralNet** [comment]
Hi, thank you for clarifying. I assumed fused_moe is an abstraction level that dispatches to the actual kernels based on some config, so for benchmarking simply the calls to the cuda kernel could be replaced inside of the triton fused_moe with calls to the new triton kernel.
Otherwise adding a separate alternative to the entire fused_moe path would involve writing an entire block under layers? Or maybe I'm not seeing something. I don't think there is a block that would let to simply import the new kernel and use it without any further integration.

(Without both fusion and quantization, the new kernel may worsen the throughput as compared to a kernel that is already quantized and fused. Unless we benchmark without quantization and fusion first.)

Btw I noticed the fused_moe is not optimal itself, because the operations are not actually fused. These are 3 separate calls and the data will have to be copied from GPU to CPU 2 times:
# First matrix multiplication (GPU kernel)
invoke_fused_moe_kernel(...)

# SiLU activation (separate op)
torch.ops._C.silu_and_mul(intermediate_cache2, intermediate_cache1.view(-1, N))

# Second matrix multiplication (GPU kernel)
invoke_fused_moe_kernel(...)
Could be written as a single kernel

**Comment by pylNeuralNet** [comment]
As for matmul_ogs, which stands for matrix mul and outer gather/scatter, I don't see a direct quick swap option in the code, it would need some more integration, I believe.
fused_moe does it like this, if I understood right:
```python
# Gather phase
offs_token = tl.load(sorted_token_ids_ptr + offs_token_id)
a_ptrs = a_ptr + (offs_token[:, None] // top_k * stride_am + 
                  offs_k[None, :] * stride_ak)

# MatMul phase
accumulator = tl.dot(a, b, acc=accumulator)

# Scatter phase  
c_ptrs = c_ptr + stride_cm * offs_token[:, None] + stride_cn * offs_cn[None, :]
tl.store(c_ptrs, accumulator, mask=c_mask)
```
The issue is routing, vLLM uses sorted_token_ids and triton _matmul_ogs uses GatherIndx and ScatterSrcIndx.

How about replacing
```python
else:
        fused_moe_kernel[grid](
```
with an ogs implementation updated with gather and scatter index buffers?
```python
    fused_moe_kernel_ogs[grid](
        hidden_states, expert_weights,
        gather_indices, scatter_indices,
        **config
    )
```
(for a quick benchmark, for production it could be added separately instead of swapping, with an option to configure either)

**Comment by snakescott** [comment]
I'm curious (and chiming in based on the `good first issue` tag): is it clear what the configurations/arguments passed to
1. `vllm/benchmarks/benchmark_serving.py` 
2. `vllm serve` (is this the same as `python -m vllm.entrypoints.openai.api_server`?)

should be? @zyongye I see you had an initial stab upthread, was that based on https://github.com/vllm-project/vllm/blob/main/benchmarks/README.md or prior experimentation or?

My understanding (but again, noob!) of what
>  the task at hand is to benchmark the kernel by itself against the existing fused moe code path in vLLM, on a variety of models and devices.

entails is
1. an experimental branch supporting `USE_TRITON_MOE_KERNEL=1`
2. Benchmarks comparing `USE_TRITON_MOE_KERNEL={0,1}`

It looks like @pylNeuralNet and @zyongye have a great start on (1), hence my curiosity on (2).

Thanks!

**Comment by zyongye** [comment]
@snakescott Hi, first of 

> ```vllm serve``` (is this the same as ```python -m vllm.entrypoints.openai.api_server```?)

I believe the answer is yes. Since I compiled vllm from source, it is more natural to launch that way.

And I think you're right on points 1 and 2. I think for comparison, we need to compare end-to-end performance with different kernels. But eventually, my understanding is that if we see good speedup, we want to replace the old kernel completely (since the kernel right now is also based on Triton) and we don't need the env var when merged. It is fully transparent to the user.

**Comment by zyongye** [comment]
Hi all, 
   Update on my end. After further investigation into the code, I think the core function we need to modify is https://github.com/vllm-project/vllm/blob/3408e471597e7a36ca79fab5fc849f4fb5576df8/vllm/model_executor/layers/fused_moe/fused_moe.py#L1230
which basically does MLP+SwiGLU+MLP and replace it with the custom kernel in ```matmul_ogs``` and ```swiglu.swiglu``` 
https://github.com/triton-lang/triton/blob/e341f7fd752abf91911acfd30e50a9bfc9a03ad5/bench/bench/bench_mlp.py#L116-L118

However I found out that the current ```matmul_ogs``` doesn't support applying expert weight when scattering back. I posted the issue [here](https://github.com/triton-lang/triton/issues/6527).

While waiting for the response, I will assemble a test function that performs ```fused_experts_impl``` with only bf16 for activation and weight. Currently it's not a easy integration. 

p.s., triton also release their [routing](https://github.com/triton-lang/triton/pull/6521) kernel, guess that's our next step. But I already envision a larger-scale change on the kernel and relative data structure. Since the current routing function are in a different function https://github.com/vllm-project/vllm/blob/3408e471597e7a36ca79fab5fc849f4fb5576df8/vllm/model_executor/layers/fused_moe/layer.py#L781

**Comment by zyongye** [comment]
After a few days of understanding and writing, with the newly dropped routing kernel, I decided to replace the entire `forward_cuda` function https://github.com/vllm-project/vllm/blob/2ef0dc53b88ded24930f10665a0575f39aef0cac/vllm/model_executor/layers/fused_moe/layer.py#L181
with the custom function, and have my preliminary working code with tests [here](https://github.com/zyongye/vllm/commit/cdbb7831b152d94c8b8f9e089f1b4e5f91704fca). The unittests seem good so far. But several issues occur:
1. The code reqire the newest triton environment (main branch) to compile, which is not compatible with the current vllm triton lock version(3.2.0). Guess we need to wait for the updates before testing with the rest of the vLLM code. 
2. The current triton kernel lacks multiplying expert weights at the end of the second MLP layer, which leads to inefficiency. 
3. The triton [SwiGLU](https://github.com/triton-lang/triton/blob/d066e15ea226dd15601b0c1e42b64f5339a1b29f/bench/triton_bench/swiglu.py#L142) kernel has precision issue, so I still use the old `silu_and_mul`
4. The input MLP tensor `w1` and `w2` need to be transposed `.transpose(-2, -1)` before used. This seems not a big deal, we can do that after loading the data. 
5. Quantization is currently not supported. Works on `bfloat16` only now. (Help needed)

I can really use some help on 3 and 5.

**Comment by pylNeuralNet** [comment]
Hey all, to be transparent, I have to postpone my participation for abt a month, so I won't be able to actively contribute to the part that I earlier volunteered to. So please go ahead with the work. If by that time there will still be something to contribute, I'll gladly help.
Great work @zyongye !

**Comment by snakescott** [comment]
I have some bandwidth next week (not much this weekend), but I think I am much more of a noob than @zyongye or @pylNeuralNet so my main contribution may be rubber ducking 🤣. My guess is feedback from committers/maintainers would be useful.

@zyongye , wdyt about standing up your work in a branch and possibly spreading it over a few commits. 4333 LOC added seems like a lot, but I imagine some of it might be copy/paste.

* For the SwiGLU precision issue, can you say more? How did it manifest/how did you notice? Is it documented in Triton?
* Similarly, for quant, do you think the issue is that the new Triton primitive doesn't support quant at feature parity with the existing vllm kernel, or is it a vllm-side issue?

I'd be happy to chat in the vllm Slack on Monday or something to run through questions live? If I'm too noob to be useful, that's fine too (fwiw, I'm getting a lot out of learning out of engaging with this ticket!).

**Comment by zyongye** [comment]
@snakescott Thanks for helping. I reorganized the commit in the [branch](https://github.com/zyongye/vllm/tree/triton_moe), now the new code I added is separated from the code ported from triton. 

- For the SwiGLU issue, I imagine it might be because of how the function performs which elements to `silu` and which element to `mul`. In the vllm implementation https://github.com/vllm-project/vllm/blob/87aaadef73543ab3e63eea933c39cee42c418e90/vllm/model_executor/layers/activation.py#L55-L63 it performs silu in the first half of the entire array of elements along the last dimension and leaves the second half unchanged. While in triton kernel ([link](https://github.com/triton-lang/triton/blob/d066e15ea226dd15601b0c1e42b64f5339a1b29f/bench/triton_bench/swiglu.py#L146)), I notice it performs silu on the **first of every other element** while linear on the **second half**. I am not sure if this is what we need. If not, we can just leave this part unchanged.
- For quant, I think triton does support through [precisionConfig](https://github.com/triton-lang/triton/blob/d066e15ea226dd15601b0c1e42b64f5339a1b29f/bench/triton_bench/matmul_ogs.py#L410). I just haven't had the time to go through it yet and want to roll out bf16 version first.

**Comment by zyongye** [comment]
Another issue I notice regarding routing [here](https://github.com/triton-lang/triton/blob/d066e15ea226dd15601b0c1e42b64f5339a1b29f/bench/tests/test_routing.py#L11)

> the reference implementation and the triton implementation do not tie-break experts the same way

So they generate the experts in a weird way. Guess that's something to keep in mind when we do the integration tests.

**Comment by snakescott** [comment]
@zhuohan123 per https://github.com/triton-lang/triton/blob/main/RELEASE.md The new kernel may not be available through an official release until 3.4.0 (looks like it missed the cutoff for 3.3.0). Are you comfortable cherry-picking in from the commit you linked (which @zyongye ~did on his branch) or is it more important for stability/etc to wait for it to be officially available?

**Comment by snakescott** [comment]
@zyongye I had some time to look through things, here's some thoughts

First, I'm going to stand up a PR on my own fork tomorrow so we can discuss/comment inline with code and not necessarily clog this issue. To your five initial points

> 1. The code reqire the newest triton environment (main branch) to compile, which is not compatible with the current vllm triton lock version(3.2.0). Guess we need to wait for the updates before testing with the rest of the vLLM code.

Agree and mentioned above. 

> 2. The current triton kernel lacks multiplying expert weights at the end of the second MLP layer, which leads to inefficiency.
Can you say more? Which line of code in vllm does this multiplication and where would you expect it to be (where is it missing) in Triton?

>3. The triton [SwiGLU](https://github.com/triton-lang/triton/blob/d066e15ea226dd15601b0c1e42b64f5339a1b29f/bench/triton_bench/swiglu.py#L142) kernel has precision issue, so I still use the old silu_and_mul
I looked at the code and agree that vllm splits the tensor in half while triton splits by taking interleaved indexes. That said, I'm confused about how/why `matmul_ogs` calls are compatible with vllm `silu_and_mul` / split approach, since this seems like it would result in different inputs than what happens in the triton codebase post activation.


>4. The input MLP tensor w1 and w2 need to be transposed .transpose(-2, -1) before used. This seems not a big deal, we can do that after loading the data.
I assume this is because `matmul_ogs` assumes the tensors come in transposed? I'm not sure of the performance implications of the `contiguous` call but would certainly guess that reworking either vllm or triton code to switch native layout of these tensors would be a non starter.

>5. Quantization is currently not supported. Works on bfloat16 only now. (Help needed)

I agree with your later point about `precisionConfig`. I sort of expect translation from vllm quant format (`use_abc_xyz` function args?) to triton format (`precisionConfig`) to be necessary? I agree with getting it to work with bfloat16 initially, although I have some ideas about refactors which might make it slightly cleaner.

I should have time to get my PR up tomorrow, and will link it. Look forward to collaborating on this!

**Comment by zyongye** [comment]
I have an experiment integration code here. And the result is just showing a comparable result on A6000 machine. 
Code link: [here](https://github.com/zyongye/vllm/commit/e75a4c415777b596bbea86ef4fda71d1d697fa23)

Setup:
CPU: Intel Xeon Gold 6126 CPU @ 2.60GHz
GPU: RTX 8xA6000, driver version 560.35.05, CUDA 12.6 (shard model with TP if needed)

Dataset: ShareGPT with 100 sequences (infqps)

Results:

![Image](https://github.com/user-attachments/assets/385918aa-9951-418a-ba4b-657e3093ccf9)


![Image](https://github.com/user-attachments/assets/3cb3f5ca-b286-4d76-9563-de2c5b61f981)


![Image](https://github.com/user-attachments/assets/35cda426-4c68-48e3-b8e7-1f464ec693e4)

The most straightforward comparison would be Qwen3 in the last bar, since that runs with the original Triton kernel without any modification (explained below). But we can see that for ttft the average improvement is about 10% and tpot is about 20-30% improvement. So I think it is worthwhile to do a full integration. 

Several things I noticed:
1. The naive triton `routing` kernel doesn't have a normalize knob, which, when is set, performing topk then softmax rather than softmax then topk. Only Qwen 3 set the renormalize to true, and I need to change the Triton kernel for other model to get them to work correctly (commit [link](https://github.com/zyongye/triton/commit/8d3d24b2ea52f161325a90f39a79ea708b27d67f)). 
2. The current torch compile (by adding `support_torch_compile`) will break the new kernel, which is disabled in all the above experiments. I may need some insight on how that flag works before I recreate a POC and submit to pytorch.
4. Since the tensor weight shape is different, I have to transpose all the weights, which incurs significant overhead when loading the weights, but since it is a one-shot thing, we can tolerate that.

TODOs:
1. At this point, I will test on a more advanced GPU (if there are somewhere I can get spot H100 with lower cost let me know) to see any improvement 
2. Implement a performant softmax scheme in the Triton kernel, 
3. Tuning some knobs (like quantization config). 

I will also need some advice on how to better integrate this into the vllm. Right now it is an environment variable, is it ok, or do we want to pass it as an engine args?

**Comment by TryinCoding** [comment]
Hi, I would love to work on this as well! Is there any way to start working on this issue or is the integration almost done?

**Comment by simon-mo** [comment]
Thank you for your interest. This integration is almost done.
